### PR TITLE
Fix remaining configuration cache compatibility issues in docs

### LINKED ISSUE #21027: Make samples and documentation snippets compatible with the configuration cache
Not all of our samples and code snippets are compatible with the configuration cache, because they predate it. For example, there are many instances of using the `Task.getProject` at the execution time. In order to announce the configuration cache as a stable feature, we should ensure that we don't promote incompatible code.

This issue is going to be delivered in several steps
- [x]  Add a (non-blocking) CI configuration to run snippets and samples tests with the configuration cache enabled
- [x] Fix common infrastructure problems (e.g. many tests fail because of extra input produced by the configuration cache)
- [ ] Fix broken samples and snippets
- [x] Exclude unsupported tests
- [x] Run the tests with CC enabled as part of the merge checks

There are certain (deprecated) Gradle features that are not going to be supported with the CC enabled. Of course, the samples/snippets covering these features are not going to be updated.

### PR DESCRIPTION
**Pull request by mlopatkin**
Fixes #21027

### CODE REVIEW

**Comment by mlopatkin** [comment]
@bot-gradle test this

**Comment by mlopatkin** [comment]
@bot-gradle test this

**Comment by mlopatkin** [comment]
@bot-gradle test and merge

File: subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/groovy/build.gradle [hunk 1]
```diff
@@ -114,9 +114,10 @@ tasks.register('packageFiles2', Zip) {
 
 // tag::adhoc-destroyable-task[]
 tasks.register('removeTempDir') {
-    destroyables.register(layout.projectDirectory.dir('tmpDir'))
+    def tempDir = layout.projectDirectory.dir('tmpDir')
+    destroyables.register(tempDir)
     doLast {
-        delete(layout.projectDirectory.dir('tmpDir'))
```
Comment (bamboo) [inline subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/groovy/build.gradle:119]: 🤔 I think we should prefer `FileOperations#delete` over APIs.

File: subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/kotlin/build.gradle.kts [hunk 2]
```diff
@@ -114,13 +114,14 @@ tasks.register<Zip>("packageFiles2") {
 
 // tag::adhoc-destroyable-task[]
 tasks.register("removeTempDir") {
-    destroyables.register(layout.projectDirectory.dir("tmpDir"))
+    val tmpDir = layout.projectDirectory.dir("tmpDir")
+    destroyables.register(tmpDir)
     doLast {
-        delete(layout.projectDirectory.dir("tmpDir"))
```
Comment (bamboo) [inline subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/kotlin/build.gradle.kts:119]: 🤔 I think we should prefer `FileOperations#delete` over APIs.

**Review by bamboo — CHANGES_REQUESTED** [review]

File: subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/groovy/build.gradle (hunk 1 above)
Comment (mlopatkin) [reply subprojects/docs/src/snippets/tasks/incrementalBuild-customTaskClass/groovy/build.gradle:119]: Thanks, I've added this to https://github.com/gradle/gradle/issues/22536#issuecomment-1342823682. It isn't really feasible prior to having getters for injectable services.

**Pull request merged**